### PR TITLE
graph layout pass for the Fused Conv+Bias+Actv Implementation

### PR DESCRIPTION
**Pull request by deven-amd**
Part 3/3 of the changes required for the fused conv+bias+activation operator

This commit builds on the changes introduced in PR #186 + PR #199

It contains code to do a POST_PLACEMENT pass over the graph to find CBA sequences and replace them with the corresponding fused OP.

Skeleton code for the CBNA and NA fusion ops is also part of this commit

----------------------------------------------------------

Currently there is only one example that shows the Fusion API in work.

```
/root/models/tutorial/images/mnist
```
Edit the `convolutional.py` in the above testcase to change the filter size from `5x5` to `3x3` (grep for `5, 5` and change to `3, 3`). This is required because MIOpen fusion currently only supports a padding size of 1. This example uses `SAME` padding and a filter of `5x5` will result in a padding size == 2.

Once you have made the above change, simply do 

to run the testcase with fusion OFF (i.e. default behavior)
- `python3 ./convolutional.py`   

to run the testcase with fusion ON
- `TF_ROCM_ENABLE_FUSION=1 python3 ./convolutional.py` 

Setting TF_CPP_MIN_VLOG_LEVEL=2 will enable VLOG messages related to Fusion Ops.

Unit testcases for fusion will be added in the near future as well

### CODE REVIEW

**Comment by whchung** [comment]
@deven-amd Could you alter the descriptor to say a little bit about known tests / applications which could leverage this PR to do kernel fusion?

Also CI shows the following error for `//tensorflow/python/keras:core_test`

```
2018-10-10 05:10:03.846710: I tensorflow/core/graph/gpu_fusion_pass.cc:305] ROCmFusionPass was not enabled!
```

**Comment by deven-amd** [comment]
The message
```
2018-10-10 05:10:03.846710: I tensorflow/core/graph/gpu_fusion_pass.cc:305] ROCmFusionPass was not enabled!
```
should not print by default...forgot to change the "vlog level" for the VLOG messages in the code to 2 (I set it to -1 to enable them for debugging, without also enabling all the other VLOG spam)

will push -force another commit with this fixed (and some other minor changes)

File: tensorflow/core/graph/gpu_fusion_pass.cc [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+  RemoveFusedNodes(fused_nodes);
+
+  // DumpGraph("After running ROCmFusionPass", &**g);
+  return true;
+}
+
+void ROCmFusionPass::InitializeFusions(std::vector<ROCmFusionOpBase*>& fusions,
+                                       Graph* g,
+                                       std::set<const Node*>& fused_nodes) {
+  fusions.push_back(
```
Comment (whchung) [inline tensorflow/core/graph/gpu_fusion_pass.cc]: @deven-amd the life-cycle of `fusions` and `fused_nodes` might need to be reviewed. it seems these 3 instances of `ROCmFusionOpXXX` might be leaking after `fusions` goes out of scope.

**Comment by deven-amd** [comment]
re the `//tensorflow/python/keras:core_test` test. 

It passes consistently (3 out of 3) for me locally with both fusion enabled and disabled. let see if the CI failure persists after my commit

File: tensorflow/core/graph/gpu_fusion_pass.cc [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+// declaration of all the constants that are used in the gpu_fusion_pass
+// namespace
+const int kVlogLevel = 2;
+
+// graph pass grouping for this fusion pass
+const OptimizationPassRegistry::Grouping kROCmFusionPassGrouping =
+    OptimizationPassRegistry::POST_PLACEMENT;
+
+// attribute name strings
+const string kAttr_T = "T";
```
Comment (whchung) [inline tensorflow/core/graph/gpu_fusion_pass.cc]: @deven-amd in order to help upstream this pass in the future, I think it's better to change these `string` to `static const char* const`. To my understanding upstream TF tends to avoid using `std::string` unless it's used to deal with underlying operating system. Since this pass is purely in TF proper we should either use `char*` or `StringPiece`.

File: tensorflow/core/graph/gpu_fusion_pass.cc [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+
+// absract base class for an individual fusion operation
+class ROCmFusionOpBase {
+ public:
+  ROCmFusionOpBase(Graph* g, std::set<const Node*>& fn, string n)
+      : graph_(g), fused_nodes_(fn), fusion_op_name_(n) {}
+
+  virtual ~ROCmFusionOpBase() {}
+
+  virtual bool DoFusion(const Node* n) = 0;
```
Comment (whchung) [inline tensorflow/core/graph/gpu_fusion_pass.cc]: @deven-amd also please use `Status` to replace `bool` so this pass could be upstreamed in the near future.

File: tensorflow/core/graph/gpu_fusion_pass.cc [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+// is this node an instance of a convolution op for which we support fusion for?
+inline bool isOpConvolution(const Node* n) {
+  return (n->type_string() == "Conv2D");
+}
+
+// is this node an instance of a bias op for which we support fusion for?
+inline bool isOpBias(const Node* n) { return (n->type_string() == "BiasAdd"); }
+
+// is this node an instance of a activation op for which we support fusion for?
+inline bool iaOpActivation(const Node* n) {
```
[the remark this answers is not in the document]
Comment (deven-amd) [reply tensorflow/core/graph/gpu_fusion_pass.cc]: will be fixed in next commit

File: tensorflow/core/graph/gpu_fusion_pass.cc (hunk 2 above)
Comment (deven-amd) [reply tensorflow/core/graph/gpu_fusion_pass.cc]: will change to `const char*` in next commit

File: tensorflow/core/graph/gpu_fusion_pass.cc (hunk 3 above)
Comment (deven-amd) [reply tensorflow/core/graph/gpu_fusion_pass.cc]: `DoFusion` is only used within this file. 

The only routine (in this file) that can be called by code outside of this file is `ROCmFusionPass::Run` and that does have a `Status` return value.

I would prefer to leave the return value of `DoFusion` as `bool` given that it signifies whehther or not the routine applied fusion to the given node. I will add a comment to the same effect.

Given this, is it ok to leave the return value of `DoFusion` as `bool`?

Comment (whchung) [reply tensorflow/core/graph/gpu_fusion_pass.cc]: thanks for the clarification. if `DoFusion` is an internal member function used solely within this module i think it's okay to leave it to return `bool`.

**Comment by whchung** [comment]
@deven-amd I was just wondering why the last round of test took so long... thanks for restarting the test
